app/models/user.rb

```diff
@@ -0,0 +1,31 @@
+class User < ActiveRecord::Base
+
+  validates :uid, :name, :oauth_token, :oauth_expires_at, presence: true
+  validates :uid, :access_token, uniqueness: true
+
+  after_create :update_access_token!
+
+  class << self
+
+    def from_omniauth(auth)
+      find_or_initialize_by(uid: auth.uid).tap do |user|
+        user.uid              = auth.uid
+        user.name             = auth.info.name
+        user.image_url        = auth.info.image
+        user.oauth_token      = auth.credentials.token
+        user.oauth_expires_at = Time.at(auth.credentials.expires_at)
+        user.save!
+      end
```

Comment:
You could also write it like this taking advantage of the `find_or_create_by!` method instead of `find_or_initialize_by` paired with `save!` (functionally they are the same). Also, no need to use `tap`, all the `find_or_X_by` methods take a block for setting attributes:

```ruby
find_or_create_by!(uid: auth.uid) do |user|
  user.uid              = auth.uid
  user.name             = auth.info.name
  user.image_url        = auth.info.image
  user.oauth_token      = auth.credentials.token
  user.oauth_expires_at = Time.at(auth.credentials.expires_at)
end
```